### ISSUE TITLE
Validation errors

### ISSUE THREAD

**Issue by Ceprey2** [czcorpus/kontext#3365]
I get validation errors, having run
`python3 scripts/validate_setup.py conf/config.xml`

> # config.xml conforms config.rng [ ERROR ]
    Extra element favicon in interleave

# <upload_cache_dir> is present [ ERROR ]
    Path /tmp/kontext-upload does not exist or is wrong type (file/directory/link)
# ...and has right permissions [ ERROR ]
    [Errno 2] No such file or directory: '/tmp/kontext-upload'
# ...and owner/group [ ERROR ]
    [Errno 2] No such file or directory: '/tmp/kontext-upload'

# Manatee path is correct (or empty and Manatee module is importable) [ ERROR ]
    Path /var/lib/manatee/registry/manatee.py does not exist or is wrong type (file/directory/link)

# ... and the directory has a proper permissions [ ERROR ]
    Permissions to /var/local/corpora/subcorp failed. Expected: 755, actual: 755

Could these errors prevent me from seeing my newly installed corpus (since I cannot see it in KonText menu)?

**Comment by tomachalek** [comment]
In case your current installation works fine otherwise, the error can be ignored (considering you e.g. do not use filter words upload on the "word list" page). We will test the script as some of its reports look strange (there's a new ticket for this #3372).

The errors above should not prevent you from adding a new corpus. To be able to see a new corpus, two conditions must be met:

1. the corpus must be available via the "corparch" plug-in (in case you use `default_corparch`, you have probably an xml file containing a list of corpora, the file is referred in the main config.xml in `kontext/plugins/corparch/file`).
2. the corpus must be set as accessible via the "auth" plug-in. In case you use `default_auth` , the authorization info is stored in Redis. Keys are like this: `corplist:user:[user ID]` (e.g. `corplist:user:2). Values are encoded JSON lists of corpora available for the user. There is a helper script `lib/plugins/default_auth/scripts/usercorp.py` which can be used to make a new corpus avail. for a user (we assume user ID = 2):
`python3 lib/plugins/default_auth/scripts/usercorp.py 2 add my_new_corpus` (other possible actions are `remove`, `remove_all` and `list`.

Also, there must be a proper Manatee "registry" file available which I assume you already have. But even without the registry file (and considering the two conditions above are met) KonText should at least show you the corpus exists.

**Comment by Ceprey2** [comment]
Thank you very much, Tomachalek, for your detailed explanations.
For the second condition, I had made the corpus available to the users 0 and 1.
But for the first condition, I don't have any folder called "corparch". The file config.xml is int kontext/conf/
Might that be a problem?
To forcr KonText to see the new corpus, I restart it as follows:

>  systemctl restart gunicorn.service
Am I done with this?

**Comment by tomachalek** [comment]
Regarding the `corparch` plug-in part in my explanation above - I should have described it in a more understandable way. The path `kontext/plugins/corparch/file` is meant to be understood as a "path" within an XML document. So in this case there should be an element `<kontext>...<plugins>...<corparch>...<file>` present in the config.xml (which itself is located in the `conf` directory).

Concerning the problem of updating a corpora setup within a running service - the restart you mentioned is perfectly OK as KonText caches a lot of configuration info and restart is the safest way to clear all the cached info. It is also possible to send the `SIGUSR1` signal to a respective Gunicorn master process (`kill -SIGUSR1 [process ID]`) and KonText will reload the configuration too. This is mainly for  KonText instances with higher load where a lot of active processes serving users can be expected at any moment. Otherwise the users probably won't notice a difference if you perform a restart.

**Comment by Ceprey2** [comment]
Thank you,
I have just verified, there is element <file> in config.xml, the corplist file is mentioned:

> <corparch>
            <module>default_corparch</module>
            <js_module>defaultCorparch</js_module>
            <file extension-by="default">/opt/kontext/conf/corplist.xml</file>
            <root_elm_path extension-by="default">/corplist</root_elm_path>
            <tag_prefix extension-by="default">+</tag_prefix>
            <max_num_hints extension-by="default">10</max_num_hints>
            <default_page_list_size extension-by="default">30</default_page_list_size>
        </corparch>
In corplist.xml I did include the new corpus id:

> <kontext>
    <corplist>
	     
	     <corpus ident="ParallKorpUkrChartUN"/>
	     <corpus ident="SUSANNE"/>

    </corplist>
</kontext>


Having restarted KonText with

> systemctl restart gunicorn.service

I still can see only SUSANNE corpus.

I did not manage to identify the process. Am I supposed to find the process ID with apps like htop?
I'm sorry, it seems strange that I can't see the new corpus. As you have mentioned, I would see it at least listed even without data. Should I probably start a new issue?

**Comment by tomachalek** [comment]
If you use the latest `master` version of KonText, enter an address like this to your browser:

```
http://[your_domain]/query?corpname=parallkorpukrchartun
```
(in case you use KonText 0.13.x, just replace "query" with "first_form").

We assume here you're logged in or the corpus is accessible for the *public* user (in config.xml, in the `auth` plug-in section, there should be the `anonymous_user_id` tag with a number specifying which user ID is for the public user).

In case the page loads with no error, authorization works and the corpus is also enabled and the problem is actually somewhere in the "corparch" plug-in which provides listing of corpora. 

I am going to briefly test the "default_corparch" plug-in to make sure it operates properly.

**Comment by Ceprey2** [comment]
Thank you!
I assumed that it would be seen to anonimous user without logging in, sorry, I did not know I should be logged in. The matter is that I can't log in, i get the error: "Incorrect username or password". I have used the two usernames and passwords from users.sample.json:

> 
[
  {
    "id": 0,
    "username": "anonymous",
    "pwd": null,
    "firstname": "Anonymous",
    "lastname": null,
    "email": null,
    "permitted_corpora": ["susanne", "ParallKorpUkrChartUN"]
  },
  {
    "id": 1,
    "username": "your_user",
    "pwd": "yourpwd",
    "firstname": "Your",
    "lastname": "Name",
    "email": "your.email@your.domain",
    "permitted_corpora": ["susanne", "ParallKorpUkrChartUN", "your_other_corpus"]
  }
]

**Comment by tomachalek** [comment]
This is strange. I've just tested the latest version from the `master` branch. I've removed all my current users from Redis and run 

```
python3 lib/plugins/default_auth/scripts/install.py lib/plugins/default_auth/scripts/users.sample.json
```
... and I can log-in without a problem (as `your_user`). Can you please post here part of your config.xml containing `corparch` and `auth` plug-ins?

**Comment by Ceprey2** [comment]
Thank you for your patience! Hereafter are the plugins you've asked for:

>         <corparch>
            <module>default_corparch</module>
            <js_module>defaultCorparch</js_module>
            <file extension-by="default">/opt/kontext/conf/corplist.xml</file>
            <root_elm_path extension-by="default">/corplist</root_elm_path>
            <tag_prefix extension-by="default">+</tag_prefix>
            <max_num_hints extension-by="default">10</max_num_hints>
            <default_page_list_size extension-by="default">30</default_page_list_size>
        </corparch>


>   <auth>
            <module>default_auth</module>
            <js_module>defaultAuth</js_module>
            <anonymous_user_id>1</anonymous_user_id>
            <auth_cookie_name>kontext_session</auth_cookie_name>
            <login_url>/user/login?continue=%s</login_url>
            <logout_url>/user/logout?continue=%s</logout_url>
        </auth>

**Comment by tomachalek** [comment]
First thing to notice - you use ID = 1 for the anonoymous user, yet the user has a defined password and the other one (ID = 0) does not have it. Here, your only non-anonymous user has no password set, but it is actually impossible to perform a login like this AFAIK.  I believe this can be the source of your problem. Please try to set `<anonymous_user_id>0</anonymous_user_id>`, restart and then log-in as `your_user` (assuming you use the user JSON config mentioned earlier).

**Comment by Ceprey2** [comment]
Thank you, I could log in with "your_user", but I still can't see but Susanne. I've also tried to introduce in the browser:

> http://10.0.3.48:8080/first_form?corpname=parallelkorpukrchartun
But I've got the message:

> No access to corpus parallelkorpukrchartun

I have also validated the installation, getting the same 6 error, nothing new.
My corplist.xml file is:

> <kontext>
    <corplist>
	     
	     <corpus ident="ParallKorpUkrChartUN"/>
	     <corpus ident="SUSANNE"/>

    </corplist>
</kontext>

**Comment by tomachalek** [comment]
Can you please open your Redis console (`redis-cli -n 1`) and post the contents of `corplist:user:0` and `corplist:user:1`?
```
get corplist:user:0
get corplist:user:1
```

**Comment by Ceprey2** [comment]
Sure, here it is:

> root@fifth-kontext-container:~# redis-cli -n 1
127.0.0.1:6379[1]> get corplist:user:0
"[\"susanne\", \"ParallKorpUkrChartUN\"]"
127.0.0.1:6379[1]> get corplist:user:1
"[\"susanne\", \"ParallKorpUkrChartUN\"]"
127.0.0.1:6379[1]>

**Comment by tomachalek** [comment]
Everything looks fine. Can you please try one more thing? In the file `lib/controller/kontext.py`, there is a method `_check_corpus_access` (https://github.com/czcorpus/kontext/blob/master/lib/controller/kontext.py#L597). Please try to add some debugging logging as shown below, restart your Gunicorn server and see (and post here please) what is printed into the log. The added logging lines are marked with `# added line` comment.

```py3
    def _check_corpus_access(self, action_name, form, action_metadata):
        """
        Args:
            action_name:
            form:
            action_metadata:

        Returns: a 2-tuple (copus id, corpus variant)
        """
        def validate_access(cn, allowed):
            if cn in allowed:
                return True, allowed[cn]
            else:
                return False, ''

        with plugins.runtime.AUTH as auth:
            allowed_corpora = auth.permitted_corpora(self.session_get('user'))
            logging.getLogger('debug').info('allowed corp: {}'.format(allowed_corpora))  # added line
            if not action_metadata['skip_corpus_init']:
                corpname, redirect = self._determine_curr_corpus(form, allowed_corpora)
                logging.getLogger('debug').info('corpname: {}, redir: {}'.format(corpname, redirect))  # added line
                has_access, variant = validate_access(corpname, allowed_corpora)
                logging.getLogger('debug').info('has access: {}, variant: {}'.format(has_access, variant))  # added line
                if has_access and redirect:
                    url_pref = self.get_mapping_url_prefix()
                    if len(url_pref) > 0:
                        url_pref = url_pref[1:]
                    raise ImmediateRedirectException(self.create_url(
                        url_pref + action_name, dict(corpname=corpname)))
                elif not has_access:
                    auth.on_forbidden_corpus(self._plugin_api, corpname, variant)
                for al_corp in form.getlist('align'):
                    al_access, al_variant = validate_access(al_corp, allowed_corpora)
                    # we cannot accept aligned corpora without access right
                    # or with different variant (from implementation reasons in this case)
                    # than the main corpus has
                    if not al_access or al_variant != variant:
                        raise AlignedCorpusForbiddenException(al_corp, al_variant)
            else:
                corpname = ''
                variant = ''
            return corpname, variant
```

**Comment by Ceprey2** [comment]
Here it is:


2020-10-21 09:07:12,271 [debug] INFO: allowed corp: {'susanne': '', 'ParallKorpUkrChartUN': '', 'ud_fused_test_a': ''}$

2020-10-21 09:07:12,271 [debug] INFO: corpname: susanne, redir: False

2020-10-21 09:07:12,271 [debug] INFO: has access: True, variant: $

**Comment by tomachalek** [comment]
Please try it also with the `ParallKorpUkrChartUN` corpus (i.e. use something like `http://10.0.3.48:8080/first_form?corpname=parallelkorpukrchartun`)

**Comment by Ceprey2** [comment]
Sure, here they are:
root@fifth-kontext-container:/opt/kontext# cat /var/log/kontext/application.log -e | grep "INFO: allowed corp"
2020-10-21 09:07:12,271 [debug] INFO: allowed corp: {'susanne': '', 'ParallKorpUkrChartUN': '', 'ud_fused_test_a': ''}$
2020-10-21 16:49:22,682 [debug] INFO: allowed corp: {'susanne': '', 'ParallKorpUkrChartUN': '', 'ud_fused_test_a': ''}$
2020-10-21 16:49:49,371 [debug] INFO: allowed corp: {'susanne': '', 'ParallKorpUkrChartUN': '', 'ud_fused_test_a': ''}$
2020-10-21 16:49:56,059 [debug] INFO: allowed corp: {'susanne': '', 'ParallKorpUkrChartUN': '', 'ud_fused_test_a': ''}$
2020-10-21 16:49:56,089 [debug] INFO: allowed corp: {'susanne': '', 'ParallKorpUkrChartUN': '', 'ud_fused_test_a': ''}$
2020-10-21 16:49:59,149 [debug] INFO: allowed corp: {'susanne': '', 'ParallKorpUkrChartUN': '', 'ud_fused_test_a': ''}$
2020-10-21 16:51:48,680 [debug] INFO: allowed corp: {'susanne': '', 'ParallKorpUkrChartUN': '', 'ud_fused_test_a': ''}$
root@fifth-kontext-container:/opt/kontext# cat /var/log/kontext/application.log | grep "INFO: corpname"
2020-10-21 09:07:12,271 [debug] INFO: corpname: susanne, redir: False
2020-10-21 16:49:22,682 [debug] INFO: corpname: susanne, redir: False
2020-10-21 16:49:49,371 [debug] INFO: corpname: parallelkorpukrchartun, redir: False
2020-10-21 16:49:56,089 [debug] INFO: corpname: parallelkorpukrchartun, redir: False
2020-10-21 16:49:59,149 [debug] INFO: corpname: parallelkorpukrchartun, redir: False
root@fifth-kontext-container:/opt/kontext# cat /var/log/kontext/application.log -e | grep "INFO: has access"
2020-10-21 09:07:12,271 [debug] INFO: has access: True, variant: $
2020-10-21 16:49:22,682 [debug] INFO: has access: True, variant: $
2020-10-21 16:49:49,371 [debug] INFO: has access: False, variant: $
2020-10-21 16:49:56,089 [debug] INFO: has access: False, variant: $
2020-10-21 16:49:59,149 [debug] INFO: has access: False, variant: $

**Comment by tomachalek** [comment]
OK, thank you. I've been finally able to resolve the issue. The problem is in case sensitivity in the corpus access testing routine. This is definitely a bug as KonText should treat corpus identifier as case insensitive everywhere in the code. I'll publish a fix today or tomorrow.

** Quick fix (method A)

update your respective Redis records, e.g.:

```
set corplist:user:1 "[\"susanne\", \"parallelkorpukrchartun\"]"

set corplist:user:0 "[\"susanne\", \"parallelkorpukrchartun\"]"
```

** A different fix (method B)

Change corpus name in the user list JSON:

```json
[
{
"id": 0,
"username": "anonymous",
"pwd": null,
"firstname": "Anonymous",
"lastname": null,
"email": null,
"permitted_corpora": ["susanne", "parallelkorpukrchartun"]
},
{
"id": 1,
"username": "your_user",
"pwd": "yourpwd",
"firstname": "Your",
"lastname": "Name",
"email": "your.email@your.domain",
"permitted_corpora": ["susanne", "parallelkorpukrchartun", "your_other_corpus"]
}
]
```
... and run a respective script.

Please note that once the fix is available, it will be possible to use mixed case strings (just like ParallKorpUkrChartUN is).

**Comment by tomachalek** [comment]
Fixed in the `master` branch: #3437

**Comment by Ceprey2** [comment]
Thanks a lot! I can definitely see it now!
